Add SPDM 1.3 new feature:get_key_pair_info

Wenxing-hou opened the pull request:
Refer the issue: #2293

jyao1 commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,266 @@
+        return LIBSPDM_STATUS_UNSUPPORTED_CAP;
+    }
+
+    if ((key_pair_id == 0) || (key_pair_id > SPDM_MAX_SLOT_COUNT)) {
```
@steven-bellock, do you agree this check: `key_pair_id > SPDM_MAX_SLOT_COUNT` ?

steven-bellock replied:
The number of `KeyPairId`s can be larger than the number of certificate slots. But we can run that by the SPDM Working Group to ensure that's what was intended in the specification.

jyao1 commented on library/spdm_responder_lib/libspdm_rsp_key_pair_info.c:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    public_key_info_len =
+        spdm_context->local_context.key_pair_info[key_pair_id].public_key_info_len;
```
why you leave 0 to be empty slot?

Wenxing-hou replied:
Thanks. I have fixed the issue. I forget that the key_pair_id starts from 1.

jyao1 commented on include/internal/libspdm_common_lib.h:
```diff
@@ -78,6 +78,17 @@ typedef struct {
+    uint16_t public_key_info_len;
+    uint8_t assoc_cert_slot_mask;
+    uint8_t public_key_info[LIBSPDM_MAX_PUBLIC_KEY_INFO_LEN];
+} libspdm_key_pair_info;
```
it should be `libspdm_key_pair_info_t`

Wenxing-hou replied:
Thanks. I have made the format consistent.

jyao1 commented on include/internal/libspdm_common_lib.h:
```diff
@@ -150,6 +165,10 @@ typedef struct {
     bool multi_key_conn_rsp;
+
+    /*responder key pair info*/
+    libspdm_key_pair_info_t key_pair_info[SPDM_MAX_SLOT_COUNT];
```
I think we need a new MACRO - `LIBSPDM_MAX_KEY_PAIR_COUNT`

Wenxing-hou replied:
Thanks. I have added the new MACRO.

jyao1 commented on include/industry_standard/spdm.h:
```diff
@@ -19,6 +19,7 @@
 #define SPDM_MAX_SLOT_COUNT 8
 #define SPDM_MAX_OPAQUE_DATA_SIZE 1024
 #define SPDM_MAX_CSR_TRACKING_TAG 7
+#define LIBSPDM_MAX_KEY_PAIR_COUNT 16
```
you need to move this to spdm_lib_config.h

Wenxing-hou replied:
Thanks. I have changed the location.

Wenxing-hou commented on include/library/spdm_lib_config.h:
```diff
@@ -205,6 +209,10 @@
 #endif
 
+#ifndef LIBSPDM_MAX_PUBLIC_KEY_INFO_LEN
+#define LIBSPDM_MAX_PUBLIC_KEY_INFO_LEN 1024
```
I have changed the size to 32.

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,290 @@
+    spdm_response = (void *)(message);
+    spdm_response_size = message_size;
+
+    libspdm_zero_mem(spdm_response, spdm_response_size);
```
Do not write to the receive buffer. It may be read-only.

Wenxing-hou replied:
Yes. I have deleted the operation according to the PR: https://github.com/DMTF/libspdm/pull/2763

steven-bellock commented:
For the Requester API I'm not sure there's much value to storing the `KEY_PAIR_INFO` info into the `spdm_context`. The Requester already gets its relevant key usage information from `GET_DIGESTS`. `KEY_PAIR_INFO` is really for `SET_KEY_PAIR_INFO`, so I could see `libspdm_get_key_pair_info` looking like
```
libspdm_get_key_pair_info(
  void *spdm_context,
  const uint32_t *session_id,
  uint8_t key_pair_id,
  uint8_t *total_key_pairs,
  uint16_t *capabilities
  .
  .
  )
```
the values returned can then be evaluated and used in `libspdm_set_key_pair_info`.

Wenxing-hou commented on include/internal/libspdm_common_lib.h:
```diff
@@ -113,6 +124,10 @@ typedef struct {
     /*The device role*/
     bool is_requester;
+
+    /*responder key pair info*/
```
Thanks. I have added the MACRO.

Wenxing-hou commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,290 @@
+    uint32_t asym_algo_capabilities;
+    uint32_t current_asym_algo;
+    uint8_t index;
+
```
Thanks. I have added the comment.

Wenxing-hou commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,290 @@
+    spdm_request->header.param2 = 0;
+    spdm_request->key_pair_id = key_pair_id;
+    spdm_request_size = sizeof(spdm_get_key_pair_info_request_t);
+    LIBSPDM_ASSERT(spdm_request->header.spdm_version >= SPDM_MESSAGE_VERSION_13);
```
Thanks. I have deleted the check. And create a new issue for other feature: https://github.com/DMTF/libspdm/issues/2777

Wenxing-hou commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,290 @@
+    spdm_response_size = sizeof(spdm_key_pair_info_response_t) + spdm_response->public_key_info_len;
+
+    /* -=[Process Response Phase]=- */
+    if (spdm_response->total_key_pairs > LIBSPDM_MAX_KEY_PAIR_COUNT) {
```
Thanks. I have deleted the check.

jyao1 commented on os_stub/spdm_device_secret_lib_sample/lib.c:
```diff
@@ -2229,3 +2229,88 @@ bool libspdm_event_get_types(
+    libspdm_key_pair_info_t key_pair_info[LIBSPDM_MAX_KEY_PAIR_COUNT];
+
+    /*provisioned key pair info*/
+    uint8_t public_key_info_rsa2048[] = {0x30, 0x0D, 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7,
```
needs to add all algorithms.

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,271 @@
+    spdm_response_size = sizeof(spdm_key_pair_info_response_t) + spdm_response->public_key_info_len;
+
+    /* -=[Process Response Phase]=- */
+    if ((spdm_response->key_pair_id != key_pair_id) ||
```
This belongs in `Validate Response Phase`.

Wenxing-hou replied:
Yes. I have moved the location.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_key_pair_info.c:
```diff
@@ -0,0 +1,152 @@
+                                               response_size, response);
+    }
+
+    key_pair_id = spdm_request->key_pair_id;
```
This check should come much later, at around line 104.

Wenxing-hou replied:
Sure. Thanks.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_key_pair_info.c:
```diff
@@ -0,0 +1,152 @@
+    public_key_info_len = (uint16_t)(*response_size - sizeof(spdm_key_pair_info_response_t));
+    libspdm_zero_mem(response, *response_size);
+
+    uint8_t *public_key_info;
```
Declare this at the top of the function.

Wenxing-hou replied:
I have fixed the code.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_key_pair_info.c:
```diff
@@ -0,0 +1,152 @@
+    result = libspdm_read_key_pair_info(
+        spdm_context,
+        key_pair_id,
+        spdm_context->connection_info.algorithm.base_asym_algo,
```
Can probably remove this parameter. Key pairs are independent of the currently negotiated algorithm.

Wenxing-hou replied:
Hi all. I have some confused. @jyao1  @steven-bellock  In the SPDM spec, 
```
This request and its response shall report information for all key pairs on the Responder
independent of any negotiated parameters of the current SPDM connection.
```

But in the `Table 103 — KEY_PAIR_INFO response message format`,
![image](https://github.com/user-attachments/assets/60013914-309a-4d1e-8321-882ba38437fc)

Do we need to check if the CurrentAsymAlgo is configured correctly?

https://github.com/DMTF/libspdm/pull/2771#discussion_r1690685506

steven-bellock replied:
>Do we need to check if the CurrentAsymAlgo is configured correctly?

Not right now. That check would go in `libspdm_check_context` and it would make sure that the algorithms in `CurrentAsymAlgo` match the ones in the Responder's local context for `BaseAsymSel`.

steven-bellock commented on include/internal/libspdm_responder_lib.h:
```diff
@@ -938,4 +938,11 @@ bool libspdm_generate_measurement_signature(libspdm_context_t *spdm_context,
+libspdm_return_t libspdm_get_response_key_pair_info(libspdm_context_t *spdm_context,
+                                                    size_t request_size, const void *request,
+                                                    size_t *response_size, void *response);
+
```
Remove whitespace.

Wenxing-hou replied:
Thanks. Fixed.

steven-bellock commented on include/hal/library/responder/key_pair_info.h:
```diff
@@ -0,0 +1,53 @@
+ * @param  current_key_usage          Indicate the currently configured key usage for the requested key pairs ID.
+ * @param  asym_algo_capabilities     Indicate the asymmetric algorithms the Responder supports for this key pair ID.
+ * @param  current_asym_algo          Indicate the currently configured asymmetric algorithm for this key pair ID..
+ * @param  public_key_info_len        On input, indicate the size in bytes of the destination buffer to store.
```
Swap position of `public_key_info_len` and `assoc_cert_slot_mask`. `public_key_info_len` and `public_key_info` should be adjacent to one another.

Wenxing-hou replied:
Thanks. The previous design is followed the SPDM spec. Now I have changed the design to follow your feedback.

![image](https://github.com/user-attachments/assets/6863d364-16f2-4104-9f27-48591a899802)

steven-bellock commented on include/industry_standard/spdm.h:
```diff
@@ -1214,6 +1224,66 @@ typedef struct {
+        SPDM_KEY_PAIR_ASYM_ALGO_CAP_ED25519 | \
+        SPDM_KEY_PAIR_ASYM_ALGO_CAP_ED448)
+
+#define SPDM_MAX_PUBLIC_KEY_INFO_LEN 32
```
This is `65535`.

Wenxing-hou replied:
In SPDM spec, the public key info is: 
```
The format of this field shall be the DER encoding of the AlgorithmIdentifier structure
in an X.509 v3 certificate.
See the "4.1.2.7. Subject Public Key Info" clauses in RFC 5280 for additional details.
```

In RFC5280 spec, the AlgorithmIdentifier structure is:
![image](https://github.com/user-attachments/assets/07e9046d-f8ad-40ad-b874-c5c7dc525f69)

 * The RSA public key info len is 15.
 * The ecp256 public key info len is 21.
 * The ecp384 public key info len is 18.
 * The ecp521 public key info len is 18.
 * The sm2 public key info len is 21.
 * The ed25519 public key info len is 7.
 * The ed448 public key info len is 7.

So the 32 is enough for SPDM_MAX_PUBLIC_KEY_INFO_LEN.

steven-bellock replied:
>So the 32 is enough for SPDM_MAX_PUBLIC_KEY_INFO_LEN.

That's fine for a `LIBSPDM_` macro but this macro is for the SPDM specification and `PublicKeyInfoLen` is two bytes. We do the same thing for https://github.com/DMTF/libspdm/blob/69278feba36eb1fc661bd21d995ee309f986283a/include/industry_standard/spdm.h#L567-L568 where a certificate chain is typically only a few KiBs.

Wenxing-hou replied:
Thanks. I have changed the SPDM_MAX_PUBLIC_KEY_INFO_LEN to 32.

steven-bellock commented on include/library/spdm_lib_config.h:
```diff
@@ -205,6 +209,12 @@
 #endif
 
+#if LIBSPDM_ENABLE_CAPABILITY_GET_KEY_PAIR_INFO_CAP
+#ifndef LIBSPDM_MAX_KEY_PAIR_COUNT
```
This macro is no longer needed.

steven-bellock commented on include/library/spdm_requester_lib.h:
```diff
@@ -370,6 +370,40 @@ libspdm_return_t libspdm_get_measurement_extension_log(void *spdm_context,
+                                           uint16_t *current_key_usage,
+                                           uint32_t *asym_algo_capabilities,
+                                           uint32_t *current_asym_algo,
+                                           uint16_t *public_key_info_len,
```
Swap position of `public_key_info_len` and `assoc_cert_slot_mask`. `public_key_info_len` and `public_key_info` should be adjacent to one another.

Wenxing-hou commented on include/hal/library/responder/key_pair_info.h:
```diff
@@ -0,0 +1,53 @@
+
+#include "hal/base.h"
+#include "internal/libspdm_lib_config.h"
+#include "library/spdm_return_status.h"
```
Thanks. I have deleted the header.

Wenxing-hou commented on library/spdm_requester_lib/libspdm_req_get_key_pair_info.c:
```diff
@@ -0,0 +1,271 @@
+        return LIBSPDM_STATUS_UNSUPPORTED_CAP;
+    }
+
+    if ((key_pair_id == 0) || (key_pair_id > LIBSPDM_MAX_KEY_PAIR_COUNT)) {
```
Yes. I have deleted the MACRO.

steven-bellock reviewed (changes requested):
I filed https://github.com/DMTF/SPDM-WG/issues/3543. There may be more checks if the Responder's `SET_KEY_PAIR_INFO_CAP == 0`.

steven-bellock commented on include/hal/library/responder/key_pair_info.h:
```diff
@@ -0,0 +1,50 @@
+ * @param  spdm_context               A pointer to the SPDM context.
+ * @param  key_pair_id                Indicate which key pair ID's information to retrieve.
+ *
+ * @param  total_key_pairs            Indicate the total number of key pairs on the responder.
```
As was clarified in https://github.com/DMTF/SPDM-WG/issues/3526#issuecomment-2233350803, `total_key_pairs` is fixed. As such it can be stored in the `spdm_context` and removed from this function.

Wenxing-hou replied:
Hi. The `total_key_pair` is still returned in the function.
Because the key pair info is provisioned by the responder device, the key pair number should be consistent.

steven-bellock replied:
We'll see what @jyao1's opinion is when he gets back.

jyao1 replied:
Can we design a new API named "libspdm_get_total_key_pairs()" ?

steven-bellock replied:
>Can we design a new API named libspdm_get_total_key_pairs() ?

If it's going to be returned in a function then it should be in `libspdm_read_key_pair_info`. What I, and the SPDM specification, is saying is that since `total_key_pairs` is fixed at the beginning of the connection and cannot change then just put it in the `spdm_context`.

jyao1 replied:
I am fine with that. We can add LIBSPDM_DATA_TOTAL_KEY_PAIRS data type.

steven-bellock replied:
Just `LIBSPDM_DATA_TOTAL_KEY_PAIRS`.

jyao1 replied:
> Just LIBSPDM_DATA_TOTAL_KEY_PAIRS.

This will cause misunderstanding. `LIBSPDM_DATA_TOTAL_KEY_PAIRS` may be interpreted as total key pairs information.

Wenxing-hou replied:
Thanks. I have fixed the feedback. Please review the comparing:
https://github.com/DMTF/libspdm/compare/6fc54f2033535b9f78a1bfe0faf519e5c9396ba3..6a04b803a95ec3406e09a6d13e81e1530012a14c

steven-bellock commented:
As was clarified in https://github.com/DMTF/SPDM-WG/issues/3543#issuecomment-2273487285 if the Responder's `SET_KEY_PAIR_INFO_CAP == 0` then `KEY_PAIR_INFO.Capabilities` must be `0` and the Requester can check that. In addition the check
```
    if ((spdm_response->capabilities & SPDM_KEY_PAIR_CAP_SHAREABLE_CAP) == 0) {
        if (!libspdm_onehot0(spdm_response->assoc_cert_slot_mask)) {
            status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
            goto receive_done;
        }
    }
```
needs to be removed.

Wenxing-hou commented on unit_test/test_spdm_requester/get_key_pair_info.c:
```diff
@@ -0,0 +1,223 @@
+}
+
+/**
+ * Test 1: message could not be sent
```
Thanks. I have moved the error test to the error_test folder.

Wenxing-hou commented on unit_test/test_spdm_requester/get_key_pair_info.c:
```diff
@@ -0,0 +1,223 @@
+}
+
+/**
+ * Test 2: Successful response to get csr
```
Thanks. Fixed.

Wenxing-hou commented on unit_test/test_spdm_responder/key_pair_info.c:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Test 1: Successful response to get key pair info with key pair id 4
+ * Expected Behavior: get a RETURN_SUCCESS return code, and correct response message size and fields
```
Thanks. Fixed.

Wenxing-hou commented:
> As was clarified in [DMTF/SPDM-WG#3543 (comment)](https://github.com/DMTF/SPDM-WG/issues/3543#issuecomment-2273487285) if the Responder's `SET_KEY_PAIR_INFO_CAP == 0` then `KEY_PAIR_INFO.Capabilities` must be `0` and the Requester can check that. In addition the check
> 
> ```
>     if ((spdm_response->capabilities & SPDM_KEY_PAIR_CAP_SHAREABLE_CAP) == 0) {
>         if (!libspdm_onehot0(spdm_response->assoc_cert_slot_mask)) {
>             status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
>             goto receive_done;
>         }
>     }
> ```
> 
> needs to be removed.

Yes. I have added the check:
```
    /*If responder doesn't support SET_KEY_PAIR_INFO_CAP,the capabilities should be 0*/
    if ((!libspdm_is_capabilities_flag_supported(
            spdm_context, true, 0,
            SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_SET_KEY_PAIR_INFO_CAP)) &&
        ((spdm_response->capabilities & SPDM_KEY_PAIR_CAP_MASK) != 0)) {
        status = LIBSPDM_STATUS_INVALID_MSG_FIELD;
        goto receive_done;
    }
```

And I have deleted the assoc_cert_slot_mask check in get_key_pair_info.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_key_pair_info.c:
```diff
@@ -0,0 +1,157 @@
+
+    total_key_pairs = spdm_context->local_context.total_key_pairs;
+    key_pair_id = spdm_request->key_pair_id;
+    if ((key_pair_id == 0) || (key_pair_id >= total_key_pairs)) {
```
`key_pair_id > total_key_pairs`

Wenxing-hou replied:
Thanks. I have fixed it.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_key_pair_info.c:
```diff
@@ -0,0 +1,157 @@
+/**
```
File should be `libspdm_rsp_get_key_pair_info.c`.

Wenxing-hou replied:
Thanks. I have changed the file name.

Wenxing-hou commented on doc/user_guide.md:
```diff
@@ -421,6 +421,12 @@ Refer to spdm_server_init() in [spdm_responder.c](https://github.com/DMTF/spdm-e
 
    1.7, if PSK is required, optionally deploy PSK Hint in the call to libspdm_start_session().
 
+   1.8, if responder can support multi key pairs, the total_key_pairs need be set.
```
Thanks. I have fixed it.

steven-bellock reviewed (approved):
Seems like the pull request is in a good place to start testing.
